contributors/design-proposals/pod-safety.md

```diff
@@ -0,0 +1,431 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod Safety, Consistency Guarantees, and Storage Implicitions
+
+@smarterclayton @bprashant
+
+October 2016
+
+## Proposal and Motivation
+
+A pod represents the finite execution of one or more related processes on the
+cluster. In order to ensure higher level consistent controllers can safely
+build on top of pods, the exact guarantees around its lifecycle on the cluster
+must be clarified, and it must be possible for higher order controllers
+and application authors to correctly reason about the lifetime of those
+processes and their access to cluster resources in a distributed computing
+environment.
+
+To run most clustered software on Kubernetes, it must be possible to guarantee
+**at most once** execution of a particular pet pod at any time on the cluster.
+This allows the controller to prevent multiple processes having access to
+shared cluster resources believing they are the same entity. When a node
+containing a pet is partitioned, the Pet Set must remain consistent (no new
+entity will be spawned) but may become unavailable (cluster no longer has
+a sufficient number of members). The Pet Set guarante must be strong enough
+for an administrator to reason about the state of the cluster by observing
+the Kubernetes API.
+
+In order to reconcile partitions, an actor (human or automated) must decide
+when the partition is unrecoverable. The actor may be informed of the failure
+in an unambiguous way (e.g. the node was destroyed by a meteor) allowing for
+certainty that the processes on that node are terminated, and thus may
+resolve the partition by deleting the node and the pods on the node.
+Alternatively, the actor may take steps to ensure the partitioned node
+cannot return to the cluster or access shared resources - this is known
+as **fencing** and is a well understood domain.
+
+This proposal covers the changes necessary to ensure:
+
+* Pet Sets can ensure **at most one** semantics for each individual pet
+* Other system components such as the node and namespace controller can
+  safely perform their responsibilities without violating that guarantee
+* An administrator or higher level controller can signal that a node
+  partition is permanent, allowing the Pet Set controller to proceed.
+* A fencing controller can take corrective action automatically to heal
```

Comment:
Healing partitions means either segmenting the minority side by various means so that they cannot affect the majority, or ensuring that the processes have halted by software or hardware means.